Userland/Services/ContentAccessServer/ConnectionFromClient.cpp

```diff
@@ -0,0 +1,289 @@
+{
+    s_connections.set(1, *this);
+}
+
+void ConnectionFromClient::die()
+{
+    s_connections.remove(client_id());
+    GUI::Application::the()->quit();
+}
+
+RefPtr<GUI::Window> ConnectionFromClient::create_dummy_child_window(i32 window_server_client_id, i32 parent_window_id)
+{
+    auto window = GUI::Window::construct();
+    window->set_opacity(0);
+    window->set_frameless(true);
+    window->set_window_mode(GUI::WindowMode::Passive);
+    auto rect = GUI::ConnectionToWindowServer::the().get_window_rect_from_client(window_server_client_id, parent_window_id);
+    window->set_rect(rect);
+    window->show();
+    GUI::ConnectionToWindowServer::the().set_window_parent_from_client(window_server_client_id, parent_window_id, window->window_id());
+
+    return window;
+}
+
+ConnectionFromClient::ProgressObject ConnectionFromClient::create_download_progress_window(i32 window_server_client_id, i32 parent_window_id)
+{
+    auto main_window = create_dummy_child_window(window_server_client_id, parent_window_id);
+    auto dialog = GUI::Window::construct(main_window);
+    dialog->set_title("Downloading Remote File..."sv);
+    dialog->set_rect(Gfx::IntRect { 0, 0, 400, 100 });
+    dialog->center_within(*main_window);
+    auto frame = MUST(dialog->set_main_widget<GUI::Widget>());
+    MUST(frame->load_from_gml(R"gmlgmlgml(
+        @GUI::Widget {
+            layout: @GUI::VerticalBoxLayout {
+                margins: [20]
+            }
+
+            @GUI::Label {
+                name: "label"
+                text_alignment: "TopLeft"
+                fixed_height: 32
+            }
+
+            @GUI::Progressbar {
+                name: "progressbar"
+                fixed_height: 28
+            }
+        }
+    )gmlgmlgml"sv));
+    frame->set_fill_with_background_color(true);
+
+    auto& progressbar = *frame->find_descendant_of_type_named<GUI::Progressbar>("progressbar");
+    auto& label = *frame->find_descendant_of_type_named<GUI::Label>("label");
+
+    dialog->show();
+
+    return {
+        .main_window = main_window.release_nonnull(),
+        .dialog = move(dialog),
+        .update = [&progressbar, &label](AK::URL const& url, u32 current, Optional<u32> total) {
+            if (!total.has_value()) {
+                label.set_text(DeprecatedString::formatted("Downloading from {}", url));
+                progressbar.set_visible(false);
+                return;
+            }
+
+            label.set_text(DeprecatedString::formatted("Downloading {} from {}", human_readable_size(*total), url));
+            progressbar.set_range(0, *total);
+            progressbar.set_value(current);
+            progressbar.set_format(GUI::Progressbar::Percentage);
+        },
+    };
+}
+
+void ConnectionFromClient::request_url_handler(i32 request_id, i32 window_server_client_id, i32 parent_window_id, URL const& url, ShouldPrompt prompt)
```

Comment:
i must tell that itʼs pretty hard to follow! it seems it could be divided a little—definitely at L145 (maybe in other places, probably also at L193).

it seems that you could extract a boolean-returning function that would just return `approved` value immediately (named e.g. `is_remote_access_allowed` or so).